Add support for WebGL2 read and draw buffer settings

mmatyas opened the pull request:
Adds support for the `ReadBuffer` and `DrawBuffers` WebGL2 calls and the related parameter getters.

See:

- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.2
- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.4
- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.11

<!-- Please describe your changes on the following line: -->

This is marked as WIP because with these functions added (but apparently not directly related to them), some tests now run into states that produces crash.

cc @jdm @zakorgy 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mmatyas committed:
Add support for WebGL2 read and draw buffer settings

Adds support for the `ReadBuffer` and `DrawBuffers`
WebGL2 calls and the related parameter getters.

See:

- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.2
- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.4
- https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.11

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webglframebuffer.rs, components/script/dom/webgl2renderingcontext.rs
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webglframebuffer.rs, components/script/dom/webgl2renderingcontext.rs

nox commented:
This needs a rebase.

mmatyas commented:
Indeed, updated.

nox commented:
Nice!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 7537add has been approved by `nox`

<!-- @bors-servo r=nox 7537add2ebe1dba83c5079cb36ff201248af5d0c -->

bors-servo commented:
:hourglass: Testing commit 7537add2ebe1dba83c5079cb36ff201248af5d0c with merge 59a64684e0afc1613e2f2ea61734177625721691...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/TYFAlVHgRZ-oyYKcytNz1g)

CYBAI commented:
Maybe need an expectation update for those new FAIL tests 👀? (there are also some CRASH tests)

<details>

```
  │   → assert_true: Property either does not exist or is not a function: texStorage3D expected true got false
  │ FAIL [expected PASS] WebGL test #7: Property either does not exist or is not a function: texSubImage3D
  │   → assert_true: Property either does not exist or is not a function: texSubImage3D expected true got false
  │ FAIL [expected PASS] WebGL test #8: Property either does not exist or is not a function: copyTexSubImage3D
  │   → assert_true: Property either does not exist or is not a function: copyTexSubImage3D expected true got false
  │ FAIL [expected PASS] WebGL test #9: Property either does not exist or is not a function: compressedTexImage3D
  │   → assert_true: Property either does not exist or is not a function: compressedTexImage3D expected true got false
  │ FAIL [expected PASS] WebGL test #10: Property either does not exist or is not a function: compressedTexSubImage3D
  │   → assert_true: Property either does not exist or is not a function: compressedTexSubImage3D expected true got false
  │ FAIL [expected PASS] WebGL test #11: Property either does not exist or is not a function: getFragDataLocation
  │   → assert_true: Property either does not exist or is not a function: getFragDataLocation expected true got false
  │ FAIL [expected PASS] WebGL test #12: Property either does not exist or is not a function: vertexAttribI4i
  │   → assert_true: Property either does not exist or is not a function: vertexAttribI4i expected true got false
  │ FAIL [expected PASS] WebGL test #13: Property either does not exist or is not a function: vertexAttribI4iv
  │   → assert_true: Property either does not exist or is not a function: vertexAttribI4iv expected true got false
  │ FAIL [expected PASS] WebGL test #14: Property either does not exist or is not a function: vertexAttribI4ui
  │   → assert_true: Property either does not exist or is not a function: vertexAttribI4ui expected true got false
  │ FAIL [expected PASS] WebGL test #15: Property either does not exist or is not a function: vertexAttribI4uiv
  │   → assert_true: Property either does not exist or is not a function: vertexAttribI4uiv expected true got false
  │ FAIL [expected PASS] WebGL test #16: Property either does not exist or is not a function: vertexAttribIPointer
  │   → assert_true: Property either does not exist or is not a function: vertexAttribIPointer expected true got false
  │ FAIL [expected PASS] WebGL test #17: Property either does not exist or is not a function: drawRangeElements
  │   → assert_true: Property either does not exist or is not a function: drawRangeElements expected true got false
  │ FAIL [expected PASS] WebGL test #18: Property either does not exist or is not a function: getIndexedParameter
  │   → assert_true: Property either does not exist or is not a function: getIndexedParameter expected true got false
  │ FAIL [expected PASS] WebGL test #19: Property either does not exist or is not a function: createVertexArray
  │   → assert_true: Property either does not exist or is not a function: createVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #20: Property either does not exist or is not a function: deleteVertexArray
  │   → assert_true: Property either does not exist or is not a function: deleteVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #21: Property either does not exist or is not a function: isVertexArray
  │   → assert_true: Property either does not exist or is not a function: isVertexArray expected true got false
  │ FAIL [expected PASS] WebGL test #22: Property either does not exist or is not a function: bindVertexArray
  │   → assert_true: Property either does not exist or is not a function: bindVertexArray expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  │ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  │ assertFunction@http://web-platform.test:8000/_webgl/conformance2/context/methods-2.html:286:17
  └ @http://web-platform.test:8000/_webgl/conformance2/context/methods-2.html:304:25
```

```
  â–¶ OK [expected ERROR] /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #5: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #11: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #17: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #23: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #29: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #35: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:132:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:64:56

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #6: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #12: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #18: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #24: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #30: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #36: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:136:19
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:64:56

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #2: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #8: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #14: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #20: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #26: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #32: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup draw framebuffer should succeed expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:132:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:63:56

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #7: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #13: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #19: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #25: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #31: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:103:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:63:56

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #3: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #9: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #15: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #21: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #27: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ FAIL [expected PASS] WebGL test #33: Framebuffer incomplete when setup draw framebuffer.
  â”‚   â†’ assert_true: Framebuffer incomplete when setup draw framebuffer. expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:136:19
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:63:56

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:
  â”‚ FAIL [expected PASS] WebGL test #4: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #10: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #16: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #22: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #28: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ FAIL [expected PASS] WebGL test #34: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed
  â”‚   â†’ assert_true: getError expected: NO_ERROR. Was INVALID_ENUM : setup read framebuffer should succeed expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ blitframebuffer_srgb_and_linear_drawbuffers@http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:103:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/blitframebuffer-srgb-and-linear-drawbuffers.html:64:56


  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/read-draw-when-missing-image.html:
  â”‚ FAIL [expected PASS] WebGL test #3: getError expected: INVALID_OPERATION. Was INVALID_FRAMEBUFFER_OPERATION : Should generate INVALID_OPERATION when reading from a color buffer without image.
  â”‚   â†’ assert_true: getError expected: INVALID_OPERATION. Was INVALID_FRAMEBUFFER_OPERATION : Should generate INVALID_OPERATION when reading from a color buffer without image. expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ read@http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:121:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:68:5

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/read-draw-when-missing-image.html:
  â”‚ FAIL [expected PASS] WebGL test #2: getError expected: INVALID_OPERATION. Was INVALID_ENUM : Should generate INVALID_OPERATION when reading from a color buffer without image.
  â”‚   â†’ assert_true: getError expected: INVALID_OPERATION. Was INVALID_ENUM : Should generate INVALID_OPERATION when reading from a color buffer without image. expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ read@http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:116:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:68:5

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/read-draw-when-missing-image.html:
  â”‚ FAIL [expected PASS] WebGL test #4: getError expected: INVALID_OPERATION. Was INVALID_ENUM : Should generate INVALID_OPERATION when reading from a color buffer without image.
  â”‚   â†’ assert_true: getError expected: INVALID_OPERATION. Was INVALID_ENUM : Should generate INVALID_OPERATION when reading from a color buffer without image. expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ read@http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:125:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/read-draw-when-missing-image.html:68:5

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/read-draw-when-missing-image.html:
  â”‚ FAIL [expected PASS] WebGL test #5: successfullyParsed should be true (of type boolean). Was undefined (of type undefined).
  â”‚   â†’ assert_true: successfullyParsed should be true (of type boolean). Was undefined (of type undefined). expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ shouldBe@http://web-platform.test:8000/_webgl/js/js-test-pre.js:406:19
  â”‚ shouldBeTrue@http://web-platform.test:8000/_webgl/js/js-test-pre.js:432:37
  â”” @http://web-platform.test:8000/_webgl/js/js-test-post.js:24:13


  â–¶ CRASH [expected ERROR] /_webgl/conformance2/rendering/fs-color-type-mismatch-color-buffer-type.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ DESCRIPTION: This test verifies that the color types of fragment shader's outputs should match color buffers' types.
  â”‚ PASS WebGL context exists
  â”‚ PASS Succeed to set up program
  â”‚ FAIL Framebuffer incomplete.
  â”‚ FAIL Framebuffer incomplete.
  â”‚ FAIL Framebuffer incomplete.
  â”‚ FAIL Framebuffer incomplete.
  â”‚ FAIL Framebuffer incomplete.
  â”‚ FAIL Framebuffer incomplete.
  â”‚ [2020-03-06T14:09:49Z ERROR canvas::webgl_thread] Last GL operation failed: InitializeFramebuffer { color: true, depth: false, stencil: false }
  â”‚ Unexpected WebGL error: 0x506 (1286) [InitializeFramebuffer { color: true, depth: false, stencil: false }] (thread WebGL thread, at components/canvas/webgl_thread.rs:2019)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: std::panicking::begin_panic_fmt
  â”‚              at src/libstd/panicking.rs:332
  â”‚    5: canvas::webgl_thread::WebGLImpl::apply
  â”‚    6: _ZN6canvas12webgl_thread11WebGLThread10handle_msg17ha6f57803bad4a4e6E.llvm.8644076800139079754
  â”‚    7: canvas::webgl_thread::WebGLThread::process
  â”‚    8: std::sys_common::backtrace::__rust_begin_short_backtrace
  â”‚    9: std::panicking::try::do_call
  â”‚   10: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚   11: core::ops::function::FnOnce::call_once{{vtable.shim}}
  â”‚   12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚   13: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚       std::sys_common::thread::start_thread
  â”‚              at src/libstd/sys_common/thread.rs:13
  â”‚       std::sys::unix::thread::Thread::new::thread_start
  â”‚              at src/libstd/sys/unix/thread.rs:80
  â”‚   14: start_thread
  â”‚   15: __clone
  â”‚ [2020-03-06T14:09:49Z ERROR servo] Unexpected WebGL error: 0x506 (1286) [InitializeFramebuffer { color: true, depth: false, stencil: false }]
  â”‚ called `Result::unwrap()` on an `Err` value: () (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/webglrenderingcontext.rs:1446)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: core::panicking::panic_fmt
  â”‚              at src/libcore/panicking.rs:85
  â”‚    5: core::option::expect_none_failed
  â”‚              at src/libcore/option.rs:1211
  â”‚    6: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::WebGLRenderingContextMethods>::DrawingBufferWidth
  â”‚    7: _ZN3std9panicking3try7do_call17h5505e43e68696d49E.llvm.8562044443237619743
  â”‚    8: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚    9: script::dom::bindings::codegen::Bindings::WebGL2RenderingContextBinding::WebGL2RenderingContextBinding::get_drawingBufferWidth
  â”‚   10: script::dom::bindings::utils::generic_call
  â”‚   11: _Z12CallJSNativeP9JSContextPFbS0_jPN2JS5ValueEEN2js10CallReasonERKNS1_8CallArgsE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:456
  â”‚       _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:548
  â”‚   12: _ZL12InternalCallP9JSContextRKN2js13AnyInvokeArgsENS1_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:617
  â”‚       _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:634
  â”‚       _ZN2js10CallGetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_NS2_13MutableHandleIS4_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:758
  â”‚   13: _ZL10CallGetterP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_5ValueEEENS2_IPN2js5ShapeEEENS1_13MutableHandleIS6_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2282
  â”‚       _ZL19GetExistingPropertyILN2js7AllowGCE1EEbP9JSContextNS0_11MaybeRootedIN2JS5ValueEXT_EE10HandleTypeENS4_IPNS0_12NativeObjectEXT_EE10HandleTypeENS4_IPNS0_5ShapeEXT_EE10HandleTypeENS7_17MutableHandleTypeE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2334
  â”‚       _ZL23NativeGetPropertyInlineILN2js7AllowGCE1EEbP9JSContextNS0_11MaybeRootedIPNS0_12NativeObjectEXT_EE10HandleTypeENS4_IN2JS5ValueEXT_EE10HandleTypeENS4_INS9_11PropertyKeyEXT_EE10HandleTypeE12IsNameLookupNSB_17MutableHandleTypeE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2594
  â”‚       _ZN2js17NativeGetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2631
  â”‚   14: _ZN2js11GetPropertyEP9JSContextN2JS6HandleIP8JSObjectEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/ObjectOperations-inl.h:117
  â”‚       _ZN2jsL25GetObjectElementOperationEP9JSContext4JSOpN2JS6HandleIP8JSObjectEENS4_INS3_5ValueEEES9_NS3_13MutableHandleIS8_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter-inl.h:486
  â”‚       _ZN2jsL19GetElementOperationEP9JSContext4JSOpN2JS6HandleINS3_5ValueEEES6_NS3_13MutableHandleIS5_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter-inl.h:600
  â”‚       _ZN2js3jit17DoGetElemFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_18ICGetElem_FallbackEN2JS6HandleINS7_5ValueEEESA_NS7_13MutableHandleIS9_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineIC.cpp:1890
  â”‚   15: <unknown>
  â”‚   16: <unknown>
  â”‚   17: _ZL13EnterBaselineP9JSContextRN2js3jit12EnterJitDataE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineJIT.cpp:113
  â”‚       _ZN2js3jit32EnterBaselineInterpreterAtBranchEP9JSContextPNS_16InterpreterFrameEPh
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineJIT.cpp:186
  â”‚   18: <unknown>
  â”‚ [2020-03-06T14:09:49Z ERROR servo] called `Result::unwrap()` on an `Err` value: ()
  â”” Pipeline failed in hard-fail mode.  Crashing!


  â–¶ CRASH [expected ERROR] /_webgl/conformance2/rendering/clear-func-buffer-type-match.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ [2020-03-06T14:10:32Z ERROR canvas::webgl_thread] Last GL operation failed: InitializeFramebuffer { color: true, depth: false, stencil: false }
  â”‚ Unexpected WebGL error: 0x506 (1286) [InitializeFramebuffer { color: true, depth: false, stencil: false }] (thread WebGL thread, at components/canvas/webgl_thread.rs:2019)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: std::panicking::begin_panic_fmt
  â”‚              at src/libstd/panicking.rs:332
  â”‚    5: canvas::webgl_thread::WebGLImpl::apply
  â”‚    6: _ZN6canvas12webgl_thread11WebGLThread10handle_msg17ha6f57803bad4a4e6E.llvm.8644076800139079754
  â”‚    7: canvas::webgl_thread::WebGLThread::process
  â”‚    8: std::sys_common::backtrace::__rust_begin_short_backtrace
  â”‚    9: std::panicking::try::do_call
  â”‚   10: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚   11: core::ops::function::FnOnce::call_once{{vtable.shim}}
  â”‚   12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚   13: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚       std::sys_common::thread::start_thread
  â”‚              at src/libstd/sys_common/thread.rs:13
  â”‚       std::sys::unix::thread::Thread::new::thread_start
  â”‚              at src/libstd/sys/unix/thread.rs:80
  â”‚   14: start_thread
  â”‚   15: __clone
  â”‚ [2020-03-06T14:10:33Z ERROR servo] Unexpected WebGL error: 0x506 (1286) [InitializeFramebuffer { color: true, depth: false, stencil: false }]
  â”‚ called `Result::unwrap()` on an `Err` value: () (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/webglrenderingcontext.rs:1446)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: core::panicking::panic_fmt
  â”‚              at src/libcore/panicking.rs:85
  â”‚    5: core::option::expect_none_failed
  â”‚              at src/libcore/option.rs:1211
  â”‚    6: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::WebGLRenderingContextMethods>::DrawingBufferWidth
  â”‚    7: _ZN3std9panicking3try7do_call17h5505e43e68696d49E.llvm.8562044443237619743
  â”‚    8: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚    9: script::dom::bindings::codegen::Bindings::WebGL2RenderingContextBinding::WebGL2RenderingContextBinding::get_drawingBufferWidth
  â”‚   10: script::dom::bindings::utils::generic_call
  â”‚   11: _Z12CallJSNativeP9JSContextPFbS0_jPN2JS5ValueEEN2js10CallReasonERKNS1_8CallArgsE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:456
  â”‚       _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:548
  â”‚   12: _ZL12InternalCallP9JSContextRKN2js13AnyInvokeArgsENS1_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:617
  â”‚       _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:634
  â”‚       _ZN2js10CallGetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_NS2_13MutableHandleIS4_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter.cpp:758
  â”‚   13: _ZL10CallGetterP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_5ValueEEENS2_IPN2js5ShapeEEENS1_13MutableHandleIS6_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2282
  â”‚       _ZL19GetExistingPropertyILN2js7AllowGCE1EEbP9JSContextNS0_11MaybeRootedIN2JS5ValueEXT_EE10HandleTypeENS4_IPNS0_12NativeObjectEXT_EE10HandleTypeENS4_IPNS0_5ShapeEXT_EE10HandleTypeENS7_17MutableHandleTypeE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2334
  â”‚       _ZL23NativeGetPropertyInlineILN2js7AllowGCE1EEbP9JSContextNS0_11MaybeRootedIPNS0_12NativeObjectEXT_EE10HandleTypeENS4_IN2JS5ValueEXT_EE10HandleTypeENS4_INS9_11PropertyKeyEXT_EE10HandleTypeE12IsNameLookupNSB_17MutableHandleTypeE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2594
  â”‚       _ZN2js17NativeGetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/NativeObject.cpp:2631
  â”‚   14: _ZN2js11GetPropertyEP9JSContextN2JS6HandleIP8JSObjectEENS3_INS2_5ValueEEENS3_INS2_11PropertyKeyEEENS2_13MutableHandleIS7_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/ObjectOperations-inl.h:117
  â”‚       _ZN2jsL25GetObjectElementOperationEP9JSContext4JSOpN2JS6HandleIP8JSObjectEENS4_INS3_5ValueEEES9_NS3_13MutableHandleIS8_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter-inl.h:486
  â”‚       _ZN2jsL19GetElementOperationEP9JSContext4JSOpN2JS6HandleINS3_5ValueEEES6_NS3_13MutableHandleIS5_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/vm/Interpreter-inl.h:600
  â”‚       _ZN2js3jit17DoGetElemFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_18ICGetElem_FallbackEN2JS6HandleINS7_5ValueEEESA_NS7_13MutableHandleIS9_EE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineIC.cpp:1890
  â”‚   15: <unknown>
  â”‚   16: <unknown>
  â”‚   17: _ZL13EnterBaselineP9JSContextRN2js3jit12EnterJitDataE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineJIT.cpp:113
  â”‚       _ZN2js3jit32EnterBaselineInterpreterAtBranchEP9JSContextPNS_16InterpreterFrameEPh
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/BaselineJIT.cpp:186
  â”‚   18: <unknown>
  â”‚ [2020-03-06T14:10:33Z ERROR servo] called `Result::unwrap()` on an `Err` value: ()
  â”” Pipeline failed in hard-fail mode.  Crashing!


  â–¶ OK [expected ERROR] /_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:
  â”‚ FAIL [expected PASS] WebGL test #2: getError expected: INVALID_OPERATION. Was NO_ERROR : Rendering to a texture where it samples from should geneates INVALID_OPERATION. Otherwise, it should be NO_ERROR
  â”‚   â†’ assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : Rendering to a texture where it samples from should geneates INVALID_OPERATION. Otherwise, it should be NO_ERROR expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ rendering_sampling_feedback_loop@http://web-platform.test:8000/_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:136:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:91:37

  â–¶ Unexpected subtest result in /_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:
  â”‚ FAIL [expected PASS] WebGL test #3: getError expected: INVALID_OPERATION. Was NO_ERROR : Rendering to a texture where it samples from should geneates INVALID_OPERATION. Otherwise, it should be NO_ERROR
  â”‚   â†’ assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : Rendering to a texture where it samples from should geneates INVALID_OPERATION. Otherwise, it should be NO_ERROR expected true got false
  â”‚ 
  â”‚ reportTestResultsToHarness/<@http://web-platform.test:8000/_webgl/js/js-test-pre.js:123:24
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1956:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ reportTestResultsToHarness@http://web-platform.test:8000/_webgl/js/js-test-pre.js:122:15
  â”‚ testFailed@http://web-platform.test:8000/_webgl/js/js-test-pre.js:249:31
  â”‚ glErrorShouldBeImpl@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1679:15
  â”‚ glErrorShouldBe@http://web-platform.test:8000/_webgl/js/webgl-test-utils.js:1653:10
  â”‚ rendering_sampling_feedback_loop@http://web-platform.test:8000/_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:136:9
  â”” @http://web-platform.test:8000/_webgl/conformance2/rendering/rendering-sampling-feedback-loop.html:92:37


  â–¶ CRASH [expected ERROR] /_webgl/conformance2/renderbuffers/readbuffer.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ 
  â”‚ [2020-03-06T14:13:42Z ERROR canvas::webgl_thread] Last GL operation failed: ReadBuffer(1029)
  â”‚ Unexpected WebGL error: 0x502 (1282) [ReadBuffer(1029)] (thread WebGL thread, at components/canvas/webgl_thread.rs:2019)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: std::panicking::begin_panic_fmt
  â”‚              at src/libstd/panicking.rs:332
  â”‚    5: canvas::webgl_thread::WebGLImpl::apply
  â”‚    6: _ZN6canvas12webgl_thread11WebGLThread10handle_msg17ha6f57803bad4a4e6E.llvm.8644076800139079754
  â”‚    7: canvas::webgl_thread::WebGLThread::process
  â”‚    8: std::sys_common::backtrace::__rust_begin_short_backtrace
  â”‚    9: std::panicking::try::do_call
  â”‚   10: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚   11: core::ops::function::FnOnce::call_once{{vtable.shim}}
  â”‚   12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚   13: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  â”‚              at /rustc/4ad62488258972bdb0e2df225d100f99ef58dbad/src/liballoc/boxed.rs:1017
  â”‚       std::sys_common::thread::start_thread
  â”‚              at src/libstd/sys_common/thread.rs:13
  â”‚       std::sys::unix::thread::Thread::new::thread_start
  â”‚              at src/libstd/sys/unix/thread.rs:80
  â”‚   14: start_thread
  â”‚   15: __clone
  â”‚ [2020-03-06T14:13:42Z ERROR servo] Unexpected WebGL error: 0x502 (1282) [ReadBuffer(1029)]
  â”‚ called `Result::unwrap()` on an `Err` value: () (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/webglrenderingcontext.rs:1446)
  â”‚ stack backtrace:
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at src/libstd/panicking.rs:474
  â”‚    3: rust_begin_unwind
  â”‚              at src/libstd/panicking.rs:378
  â”‚    4: core::panicking::panic_fmt
  â”‚              at src/libcore/panicking.rs:85
  â”‚    5: core::option::expect_none_failed
  â”‚              at src/libcore/option.rs:1211
  â”‚    6: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::WebGLRenderingContextMethods>::DrawingBufferWidth
  â”‚    7: _ZN3std9panicking3try7do_call17h5505e43e68696d49E.llvm.8562044443237619743
  â”‚    8: __rust_maybe_catch_panic
  â”‚              at src/libpanic_unwind/lib.rs:86
  â”‚    9: script::dom::bindings::codegen::Bindings::WebGL2RenderingContextBinding::WebGL2RenderingContextBinding::get_drawingBufferWidth
  â”‚   10: script::dom::bindings::utils::generic_call
  â”‚   11: _ZN2js3jit16CallNativeGetterEP9JSContextN2JS6HandleIP10JSFunctionEENS4_IP8JSObjectEENS3_13MutableHandleINS3_5ValueEEE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/VMFunctions.cpp:1559
  â”‚   12: <unknown>
  â”‚   13: <unknown>
  â”‚   14: <unknown>
  â”‚   15: <unknown>
  â”‚   16: <unknown>
  â”‚   17: _ZL8EnterJitP9JSContextRN2js8RunStateEPh
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/Jit.cpp:109
  â”‚       _ZN2js3jit13MaybeEnterJitEP9JSContextRNS_8RunStateE
  â”‚              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/aabcc9b/mozjs/js/src/jit/Jit.cpp:200
  â”‚   20: <unknown>
  â”‚ [2020-03-06T14:13:42Z ERROR servo] called `Result::unwrap()` on an `Err` value: ()
  â”” Pipeline failed in hard-fail mode.  Crashing!
```

</details>

jdm commented:
I suspect that was what was meant in the original description:
```
This is marked as WIP because with these functions added (but apparently not directly related to them), some tests now run into states that produces crash.
```
Removing the WIP marker and merging it may not have been the optimal choice :)

nox commented:
Ah sorry, my eyes saw the stock PR template and so my brain just skipped it.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25915) made this pull request unmergeable. Please resolve the merge conflicts.

mmatyas commented:
I was looked into the crashes, and when the drawbuffers method is called I get a panic from `validate_framebuffer`:

```
 0:04.08 pid:20179 [2020-03-12T11:03:56Z ERROR canvas::webgl_thread] Last GL operation failed: InitializeFramebuffer { color: true, depth: false, stencil: false }
 0:04.08 pid:20179 Unexpected WebGL error: 0x506 (1286) [InitializeFramebuffer { color: true, depth: false, stencil: false }] (thread WebGL thread, at components/canvas/webgl_thread.rs:2025)
```

(ie. `INVALID_FRAMEBUFFER_OPERATION`). I'm not really sure what's the cause of that, do you have some ideas?

At first the crash happened in `gl.clear` so I though it might be related to a missing check there. I've implemented this check, but in a separate commit as it seems this isn't actually related to the crash, just that was the first place `validate_framebuffer` was called in one of the tests. If it's indeed unrelated I can move it to a separate pull request.

jdm commented:
I looked into fs-color-type-mismatch-color-buffer-type.html and the results make sense to me - the test is relying on creating textures that have different internal formats and formats (per the spec text just above https://www.khronos.org/registry/webgl/specs/latest/2.0/#TEXTURE_TYPES_FORMATS_FROM_DOM_ELEMENTS_TABLE), and our texture validation code rejects that (as well as newer formats like RGBA8UI). The reason this causes a panic is that the test relies on initializing renderbuffers with these textures, and then the first attempt to call drawBuffers with a renderbuffer's color attachment attempts to clear the framebuffer, but since an incomplete renderbuffer is backing the framebuffer we hit a GL error when trying to perform the clear.

In short, I'm filing some issues and I think it's fine to mark the tests as expected to crash. Fixing those issues is unrelated to the work in this PR.

jdm commented:
I filed #25952 and 25954.

mmatyas commented:
Ok, rebased, added the tests and removed the clean buffer and readpixels related stuff (will send them separately). Interestingly, after the rebase the previous crashes don't stop Servo, just cause timeouts. `blitframebuffer-srgb-and-linear-drawbuffers.html` also produced some new results, but the errors seem to come from `texImage2D`.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 0afe27ef18e1bb458863bd881c79d0fa63d9c4a4 with merge ad89c2758061aa46797b58e7fa35528a80d13141...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/d7k9d4pOQOypX-zfAx5TwA)
State: approved= try=True

jdm commented:
@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 0afe27e has been approved by `nox`

<!-- @bors-servo r=nox 0afe27ef18e1bb458863bd881c79d0fa63d9c4a4 -->

bors-servo commented:
:hourglass: Testing commit 0afe27ef18e1bb458863bd881c79d0fa63d9c4a4 with merge 654dd048a3fe4578883ce38616199f06cf850111...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KwlkKxQkRIO0mka87euOLA)
Approved by: nox
Pushing 654dd048a3fe4578883ce38616199f06cf850111 to master...
